_posts/2018-12-07-fast-starting-clojure-lambdas-using-graalvm.md

```diff
@@ -0,0 +1,80 @@
+---
+layout: post
+title: Fast starting Clojure AWS Lambdas using GraalVM and a Lambda custom runtime  
+author: hjhamala, markkuko
+excerpt: Clojure suffers of a slow start up time which makes using it problematic for APIs running in AWS Lambdas. Compiling Clojure to a native binary using GraalVM and running it in a Lambda custom runtime solves the start up problem. 
+tags:
+- AWS
+- Clojure
+- Serverless 
+- Lambda
+- GraalVM
+---
+
+[**Clojure**](https://clojure.org/) is a dynamic Lisp language which is compiled to JVM bytecode for running as a normal JVM application. Starting a Clojure program is pretty slow compared to many other languages. The start up takes at least one second but depending on a program size this could be almost ten seconds.
+
+**JVM** itself starts [fast](http://clojure-goes-fast.com/blog/clojures-slow-start/) in 50 ms. The slow start up time is mostly caused by JVM class loading. Unfortunately Clojure generates a lot of classes because every Clojure variable definition and function are compiled to classes. This applies also to anonymous functions which are quite common in Clojure applications. For example a Clojure REST api with 2800 lines of code is compiled to over 200 JVM classes. The start up time of the program is almost eight seconds when running in MacBook Pro 2017 model. The minimum start up of one second or more makes Clojure also a bad choice for command line tools which are expected to run very fast.
+
+**Lambda** is a AWS serverless tecnology which lets users to run code without managing servers. Lambdas are paid only for the compute time which is consumed which makes them very attractive option for rarely used web applications. Lambdas itself have a start up time which depends on whether the Lambda application is _cold_ or _warm_. When a Lambda is first time used or there has been about 15 to 30 minutes between the last Lambda usage it takes about 600 ms to start the Lambda. After that Lambda is _warm_ and subsequent invocations are quite similar compared to applications running in normal virtual machines or containers. Performance of a Lambda depends on a memory allocated for it on creation. More memory gives also more computation time.
+
+The start up time is not a problem for Lambdas which are run as cron typed tasks, SQS queue pollers or otherwise where there is no need for quick synchronous responses. 
```

Comment:
```suggestion
The start-up time is not a problem for Lambdas which are run as cron-like tasks, SQS queue pollers or otherwise where there is no need for quick synchronous responses. 
```